### DISCUSSION TITLE
Improving support for Prometheus exemplars

### DISCUSSION THREAD

**Discussion by K-Phoen** [grafana/grafana#39930] [Dashboards, Panels & Visualizations]
Hey,

I spent some time looking into the [exemplars support in Grafana](https://grafana.com/docs/grafana/latest/basics/exemplars/), with the hope of benefiting from this extremely useful feature in my organization.

While the exemplars support is easy to get up and running, two limitations quickly stood out:
* when a query returns several dimensions (labels), the legend can be used to display only one of these dimensions on the graph. However, this selection doesn't apply to exemplars. Displaying only the exemplars related to the selected dimension is supported in the Prometheus web UI.
* Grafana samples out some exemplars to avoid cluttering the graph. I'm guessing it's primarily to avoid cluttering the view too much, but when combined with the previous point it makes drilling down and discovering interesting exemplars quite hard (and often not possible at all since the sampling process doesn't seem to distinguish between dimensions)

Here is an example of the behavior I describe in Grafana (v8.1.5):

https://user-images.githubusercontent.com/66958/135774514-873a3c9e-1e3d-4a60-8fc9-7c35fe2f51d4.mp4

And on this one, the behavior I would expect from Grafana (as seen in Prometheus v2.30.0):

https://user-images.githubusercontent.com/66958/135775676-85c5d1f1-559b-4650-87c9-f874aee9e729.mp4

Not sure whether the behavior I'm describing is a feature request or a bug so I'd love to hear what's the community and contributor's opinion :)
In any case, I think that displaying the exemplars based on what's actually being visualized would be extremely valuable. WDYT?

**Comment by aengusrooneygrafana** [comment]
@bboreham  i'll discuss this with you to understand if this is a known bug, or perhaps an issue that has been fixed now?

**Comment by bboreham** [comment]
Previous discussion of sampling: https://github.com/grafana/grafana/discussions/41972

I’m not aware that sub-selecting exemplars via the legend has been discussed before. Seems a reasonable request.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
